borg/archiver.py

```diff
@@ -406,64 +406,127 @@ def item_is_hardlink_master(item):
-            if count is None:
-                return "<deleted>"
-            return format_file_size(count)
+            if count > 0:
+                sign = '+'
+            elif count < 0:
+                sign = '-'
+            else:
+                sign = ''
+
+            return '{}{}'.format(sign, format_file_size(abs(count)))
 
         def fetch_and_compare_chunks(chunk_ids1, chunk_ids2, archive1, archive2):
             chunks1 = archive1.pipeline.fetch_many(chunk_ids1)
             chunks2 = archive2.pipeline.fetch_many(chunk_ids2)
             return self.compare_chunk_contents(chunks1, chunks2)
 
+        def sum_chunk_size(item):
+            if item.get(b'deleted'):
+                return None
+            else:
+                return sum(c[1] for c in item[b'chunks'])
+
         def get_owner(item):
             if args.numeric_owner:
                 return item[b'uid'], item[b'gid']
             else:
                 return item[b'user'], item[b'group']
 
+        def get_mode(item):
+            mode = stat.filemode(item[b'mode'])
+
+            # Special notation for hard links
+            if mode[0] == '-' and b'source' in item:
+                mode = "h" + mode[1:]
+
+            return mode
+
+        def describe_link(item):
+            if get_mode(item)[0] == 'l':
+                return 'symlink'
+            else:
+                return 'hard link'
+
+        def compare_link(item1, item2):
+            # These are the simple link cases. For special cases, e.g. if a
+            # regular file is replaced with a link, or a symlink changes to a
+            # hard link, it is indicated in compare_mode instead.
+            if b'source' not in item1:
+                if item1.get(b'deleted'):
+                    return 'added {}'.format(describe_link(item2))
+            elif b'source' not in item2:
+                if item2.get(b'deleted'):
+                    return 'removed {}'.format(describe_link(item1))
+            elif (item1[b'source'] != item2[b'source'] and
+                  get_mode(item1)[0] == get_mode(item2)[0]):
+                return 'link changed'
+
+        def compare_content(item1, item2):
+            if (not can_compare_chunk_ids or
+                    item1[b'chunks'] != item2[b'chunks']):
+                chunk_ids1 = [c[0] for c in item1[b'chunks']]
+                chunk_ids2 = [c[0] for c in item2[b'chunks']]
+                chunk_id_set1 = set(chunk_ids1)
+                chunk_id_set2 = set(chunk_ids2)
+                total1 = sum_chunk_size(item1)
+                total2 = sum_chunk_size(item2)
+
+                if (can_compare_chunk_ids or total1 != total2 or
+                    not fetch_and_compare_chunks(chunk_ids1, chunk_ids2,
+                                                    archive1, archive2)):
+                    added = sum(c[1] for c in (chunk_id_set2 - chunk_id_set1))
+                    removed = -sum(c[1] for c in (chunk_id_set1 - chunk_id_set2))
```

Comment:
(Didn't receive a notification about the mention)

can_compare_chunk_ids tells us whether the chunker params are the same. We end up in this branch in two cases:

1. We can compare chunk IDs, and the chunk ID lists were different
2. We can't compare chunk IDs (*we do not know whether the files are different*)

This second `if` is a continuation of these two branches

1. Either we hit branch (1) above (can compare, were different) then added/remove info is calculated
2. Or we hit branch (2): check if the file sizes are the same. If they aren't something must have changed. If they are the same, compare file contents.

You inverted the logic from the old commit, though. I think this needs an `else: return <nothing changed>`:

* Can't compare chunk ids
* So we hit branch (2) in the first if
* Now we do check 2, if the files are the same this would still print something like `[12345 12345]`